File: conteudo/u02 Estruturas de dados básicas lineares/Lista.java
class Lista {
    int[] array;
    int n;

    Lista() { this(6); }
    Lista(int tamanho) {
        array = new int[tamanho];
        n = 0;
    }
    
    void inserirInicio(int x) throws Exception {
        if (n >= array.length)
            throw new Exception("Erro!");

        for (int i = n; i > 0; i--) {
            array[i] = array[i - 1];
        }
        array[0] = x;
        n++;
    }

    void inserirFim(int x) throws Exception {
        if (n >= array.length)
            throw new Exception("Erro!");
        array[n] = x;
        n++;
    }

    void inserir(int x, int pos) throws Exception {
        if (n >= array.length || pos < 0 || pos > n)
            throw new Exception("Erro!");

        for (int i = n; i > pos; i--) {
            array[i] = array[i - 1];
        }
        array[pos] = x;
        n++;
    }

    int removerInicio() throws Exception {
        if (n == 0)
            throw new Exception("Erro!");
        
        int resp = array[0];
        n--;

        for (int i = 0; i < n; i++) {
            array[i] = array[i + 1];
        }

        return resp;
    }

    int removerFim() throws Exception {
        if (n == 0)
            throw new Exception("Erro!");

        return array[--n];
    }

    int remover(int pos) throws Exception {
        if (n == 0 || pos < 0 || pos >= n)
            throw new Exception("Erro!");

        int resp = array[pos];
        n--;

        for (int i = pos; i < n; i++) {
            array[i] = array[i + 1];
        }

        return resp;
    }
    
    void mostrar() {
        System.out.print("[ ");
        for (int i = 0; i < n; i++) {
            System.out.print(array[i] + " ");
        }
        System.out.println("]");
    }

    public static void main(String[] args) {
        try {
            Lista lista = new Lista();
            lista.inserirFim(1);
            lista.inserirFim(2);
            lista.inserirFim(3);
            lista.inserirInicio(0);
            lista.inserir(4, 2);
            lista.mostrar();

            System.out.println("Removido do inicio: " + lista.removerInicio());
            System.out.println("Removido do fim: " + lista.removerFim());
            System.out.println("Removido da posicao 1: " + lista.remover(1));
            lista.mostrar();
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    }
}

File: conteudo/u02 Estruturas de dados básicas lineares/Lista.c
#include <stdio.h>
#include <stdlib.h>

#define MAXTAM 100

int array[MAXTAM];
int n;

void start() {
	n = 0;
}

void inserirInicio(int x) {
    if (n >= MAXTAM)
        exit(1);

    for (int i = n; i > 0; i--) {
        array[i] = array[i - 1];
    }

    array[0] = x;
    n++;
}

void inserirFim(int x) {
    if (n >= MAXTAM)
        exit(1);

    array[n] = x;
    n++;
}

void inserir(int x, int pos) {
    if (n >= MAXTAM || pos < 0 || pos > n)
        exit(1);

    for (int i = n; i > pos; i--) {
        array[i] = array[i - 1];
    }

    array[pos] = x;
    n++;
}

int removerInicio() {
    if (n == 0)
        exit(1);

    int resp = array[0];
    n--;

    for (int i = 0; i < n; i++) {
        array[i] = array[i + 1]; 
    }

    return resp;
}

int removerFim() {
    if (n == 0)
        exit(1);

    return array[--n];
}

int remover(int pos) {
    if (n == 0 || pos < 0 || pos >= n)
        exit(1);

    int resp = array[pos];
    n--;

    for (int i = pos; i < n; i++) {
        array[i] = array[i + 1];
    }

    return resp;
}

void mostrar() {
    printf("[ ");
    for (int i = 0; i < n; i++) {
        printf("%d ", array[i]);
    }
    printf("]");
}
     
int main() {
    start();
    inserirFim(1);
    inserirFim(2);
    inserirFim(3);
    inserirInicio(0);
    inserir(4, 2);

    mostrar();
    printf("\n");

    printf("Removido do inicio: %d\n", removerInicio());
    printf("Removido do fim: %d\n", removerFim());
    printf("Removido da posicao 1: %d\n", remover(1));

    mostrar();
    printf("\n");

    return 0;
}